### PR TITLE
Infer Criteria from the current ManagedArtifact.

### PR DESCRIPTION
**Pull request by mach6**
- Removed userId from interface Criteria.
- Moved DefaultCriteria into its own file.
- DefaultManagedArtifact implements ManagedArtifact<DefaultCriteria>.
- Requested Criteria is processed from HTTP path by DefaultManagedArtifact.
  This was earlier processed by DownloadResponder.
- Removed 'managedCriteria' configuration from SeLionConfig.json.
- Revised interface ManagedArtifact.
- Interface ServerRepository updated to process HTTP path information instead of Criteria.
- Updated unit tests and documentations.

### CODE REVIEW

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/Criteria.java [hunk 1]
```diff
@@ -34,14 +34,6 @@
     String getArtifactName();
 
     /**
-     * Returns the user Id this artifact is tied to. This is essentially the folder by the name of the user that houses
-     * the artifact.
-     * 
-     * @return User Id of the criteria.
-     */
-    String getUserId();
```
Comment (mach6) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/Criteria.java]: The javadoc for the class is out of date due to this change.

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/UploadRequestProcessor.java [hunk 2]
```diff
@@ -53,7 +53,7 @@
     /**
      * @return a {@link List} of {@link ManagedArtifact} which represent items on the {@link ManagedArtifactRepository}
```
Comment (mach6) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/UploadRequestProcessor.java]: @shelar javadoc needs update as a result of this change

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultCriteria.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    private String artifactName;
+
+    private String userId;
+
+    private String applicationFolder;
+
+    public DefaultCriteria(EnumMap<RequestHeaders, String> parametersMap) {
+        validateParametersMap(parametersMap);
+        this.artifactName = parametersMap.get(RequestHeaders.FILENAME);
+        this.userId = parametersMap.get(RequestHeaders.USERID);
```
Comment (mach6) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultCriteria.java]: @shelar We seem have made the criteria for the artifact more flexible from a storage and download perspective but still have a limitation on the upload front. Case in point, see enum `UploadRequestProcessor#RequestHeaders`. Do you agree?

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/ServerRepository.java [hunk 4]
```diff
@@ -26,7 +26,7 @@
  * @param <U>
```
Comment (mach6) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/ServerRepository.java]: @shelar this javadoc is out of date due to this change set

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/Criteria.java (hunk 1 above)
Comment (mach6) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/Criteria.java]: @shelar After further thought, I think we should revert this change. A `userId` is a basic condition that a `ManagedArtifact` must conform to. It allows for the server repository to be multi-tenant.

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/ManagedArtifact.java [hunk 5]
```diff
@@ -57,7 +57,7 @@
      *            Instance of {@link Criteria} to match.
      * @return true if there is a match, false otherwise.
      */
-    <U extends Criteria> boolean matches(U criteria);
+    boolean matchesCriteria(String pathInfo);
```
Comment (mach6) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/ManagedArtifact.java]: @shelar I think this should be called `matchesPathInfo`

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/ManagedArtifact.java [hunk 6]
```diff
@@ -20,7 +20,7 @@
  * call.
  * 
  */
-public interface ManagedArtifact {
+public interface ManagedArtifact<U extends Criteria> {
```
Comment (mach6) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/ManagedArtifact.java]: @shelar With the other changes you have made, I think we should be able to avoid turning this into a template type

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultCriteria.java (hunk 3 above)
Comment (squash-merge) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultCriteria.java]: I agree, all upload side is limited by headers. I see how it defeats the purpose of the changes I submitted. Maybe, I should have modified `UploadedArtifact` to contain all header name-values in a property, and let `ServerRepository#saveContents` use it.

**Comment by mach6** [comment]
@shelar @sebady @sundaramrajendran @kumaravel-jayakumar Please take a look at this PR. Thanks!

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultCriteria.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    private String artifactName;
+
+    private String uuidFolderName;
+
+    private String subFolderName;
+
+    public DefaultCriteria(Map<String, String> parametersMap) {
+        validateParametersMap(parametersMap);
+        this.artifactName = parametersMap.get(ManagedArtifact.ARTIFACT_FILE_NAME);
+        this.uuidFolderName = parametersMap.get(DefaultRequestParameters.USER_ID);
```
Comment (sebady) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultCriteria.java]: is uuid field a uuid or a "user id"?

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultManagedArtifact.java [hunk 8]
```diff
@@ -48,29 +47,95 @@
...
+        public boolean isRequired(String parameter) {
+            if (getParameters().containsKey(parameter)) {
+                return getParameters().get(parameter).booleanValue();
+            }
+            return false;
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    public <T extends RequestParameters> T getRequstParameters() {
```
Comment (squash-merge) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultManagedArtifact.java]: Should `getRequstParameters` be changed to `getRequestParameters`?

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultCriteria.java [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+    public String getArtifactFolder() {
+        return subFolderName;
+    }
+
+    public Map<String, String> asMap() {
+        LOGGER.entering();
+        Map<String, String> contentMap = new HashMap<>();
+        contentMap.put(ManagedArtifact.ARTIFACT_FILE_NAME, getArtifactName());
+        contentMap.put(DefaultRequestParameters.USER_ID, getUUIDFolderName());
+        if (!StringUtils.isBlank(getArtifactFolder())) {
```
Comment (sebady) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultCriteria.java]: This if test is not really necessary

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultCriteria.java [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+        }
+        if (!getUUIDFolderName().equals(otherCriteria.getUUIDFolderName())) {
+            return false;
+        }
+        boolean equals = getArtifactFolder() == null ? otherCriteria.getArtifactFolder() == null
+                : getArtifactFolder().equals(otherCriteria.getArtifactFolder());
+        if (equals == false) {
+            return false;
+        }
+        return true;
```
Comment (sebady) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultCriteria.java]: Why not just return equals?

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultManagedArtifact.java [hunk 11]
```diff
@@ -48,29 +47,95 @@
...
 
-    private String parentFolderName = null;
+    private String uuidFolderName = null;
 
     private byte[] contents = null;
 
-    private final long timeToLiveInMillis;
+    private static final long timeToLiveInMillis = ConfigParser.parse().getLong(EXPIRY_CONFIG_PROPERTY);
+
+    private String requestPathInfo;
```
Comment (sebady) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultManagedArtifact.java]: Please add a comment that makes it clearer what this field  represents.  Is this the same as the full path?

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultCriteria.java (hunk 7 above)
Comment (mach6) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultCriteria.java]: Well, it's really a "uid".. Nothing about the intent requires the identifier to be associated with a user. It just needs to be a unique identifier. Probably the request parameter should be changed to reflect this too. I left that part in tact, as you can see.

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultManagedArtifact.java (hunk 8 above)
Comment (mach6) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultManagedArtifact.java]: good catch. will fix

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultCriteria.java (hunk 10 above)
Comment (mach6) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultCriteria.java]: This was pre-existing and I believe the intent was to check for equality on two instances of `DefaultCriteria` which are pointing to the same meta info. Honestly, since `Criteria` only come into play for download means, I don't think this is needed.

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultCriteria.java (hunk 9 above)
Comment (mach6) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultCriteria.java]: This was pre-existing and I believe the if check is necessary, since `artifactFolder` is an optional parameter in the `DefaultManagedArtifact` implementation, though I need to see how `asMap()` is used.

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultManagedArtifact.java (hunk 11 above)
Comment (mach6) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultManagedArtifact.java]: Will do. It's the requested path to the artifact at download time. It is not the file system absolute path.

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultManagedArtifact.java [hunk 12]
```diff
@@ -48,29 +47,95 @@
...
+    }
+
+    public void initFromPath(String pathName) {
+        // do not allow path that are empty or try to break out of the repository folder
+        Preconditions.checkArgument(StringUtils.isNotBlank(pathName));
+        Preconditions.checkArgument(!StringUtils.contains(pathName, ".."));
+
+        String repoPath = ManagedArtifactRepository.getInstance().getRepositoryFolder().getAbsolutePath();
+        String filePath = pathName;
+        if (!pathName.contains(repoPath)) {
```
Comment (mach6) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultManagedArtifact.java]: This check can fail on windows where the `repoPath` is on a different drive than the drive included in `pathName`.. Will fix this.

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/ManagedArtifact.java [hunk 13]
```diff
@@ -30,37 +70,39 @@
...
 
     /**
-     * Returns the name of the folder this artifact is housed.
+     * Returns the name of the folder this artifact is housed. This folder is relative to the {@link ServerRepository}
+     * root folder
      * 
      * @return Folder name of this artifact
      */
-    String getFolderName();
+    String getArtifactFolderName();
```
Comment (mach6) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/ManagedArtifact.java]: can be removed from the interface. can be deduced from `getPath()`. no more callers outside of `@Test`s also

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultManagedArtifact.java [hunk 14]
```diff
@@ -48,29 +49,90 @@
...
 
-    public DefaultManagedArtifact(String pathName) {
-        this.filePath = pathName;
-        artifactFile = new File(this.filePath);
-        timeToLiveInMillis = ConfigParser.parse().getLong(EXPIRY_CONFIG_PROPERTY);
+    static final class DefaultRequestParameters implements RequestParameters {
+        public static final String UID = "uid";
+
+        public Map<String, Boolean> getParameters() {
+            Map<String, Boolean> headers = new HashMap<>();
```
Comment (squash-merge) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultManagedArtifact.java]: We can avoid creating a new HashMap on every call here.

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/UploadRequestProcessor.java [hunk 15]
```diff
@@ -295,40 +289,38 @@ private int parseRequestAsFileItems() throws FileUploadException {
         }
 
         private void populateHeadersMap() {
-            EnumMap<RequestHeaders, String> headersMap = getMapFromHeaders();
+            Map<String, String> headersMap = getRequestHeadersMap();
```
Comment (squash-merge) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/UploadRequestProcessor.java:297]: SInce an ManagedArtifact's `instance` is created on the next line, could we change this code to use a single `instance`? 

```
ManagedArtifact instance = getManagedArtifactInstance();
Map<String, String> headersMap = getRequestHeadersMap(instance);
```

Corresponding modification to `getRequestHeadersMap()` would then be required.

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultManagedArtifact.java [hunk 16]
```diff
@@ -185,109 +250,11 @@ private void readContents() {
...
-            result = 31 * result + (this.getApplicationFolder() != null ? this.getApplicationFolder().hashCode() : 0);
-            return result;
-        }
-
-        @Override
-        public String toString() {
-            return "[ artifactName: " + getArtifactName() + ", userId: " + getUserId() + ", applicationFolder: "
-                    + getApplicationFolder() != null ? getApplicationFolder() : "" + " ]";
+    public String getAbsolutePath() {
+        if (artifactFile == null) {
```
Comment (squash-merge) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultManagedArtifact.java]: This 4 lines be made into one line.

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultManagedArtifact.java [hunk 17]
```diff
@@ -100,38 +177,26 @@ public String getParentFolderName() {
...
-            LOGGER.exiting(false);
-            return false;
-        }
-        if (isApplicationFolderRequested(criteria) && applicationFolderAndUserIdMatches(criteria)) {
-            LOGGER.exiting(true);
-            return true;
-        }
-        boolean matches = !isApplicationFolderRequested(criteria) && userIdMatches(criteria);
+    public boolean matchesPathInfo(String pathInfo) {
+        LOGGER.entering();
```
Comment (squash-merge) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/DefaultManagedArtifact.java]: `LOGGER.entering(pathInfo)` could be useful.

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/ManagedArtifactRepository.java [hunk 18]
```diff
@@ -120,43 +113,15 @@ public ManagedArtifact saveContents(UploadedArtifact uploadedArtifact) {
...
+    public ManagedArtifact getArtifact(String pathInfo) {
         /*
          * This method does not guarantee the presence of an artifact after isArtifactPresent() is called on an
          * artifact. This is because the cleaner thread could have deleted a expired artifact in between the calls.
          */
         try {
             repositorySynchronizationLock.lock();
-            LOGGER.entering(requestedCriteria);
-            ManagedArtifact managedArtifact = getMatch(requestedCriteria);
+            LOGGER.entering();
```
Comment (squash-merge) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/ManagedArtifactRepository.java]: `LOGGER.entering(pathInfo)` would be useful.

**Comment by squash-merge** [comment]
Thats all from me.  @sebady @sundaramrajendran  @kumaravel-jayakumar Your comments please!

File: server/src/main/java/com/paypal/selion/grid/servlets/transfer/UploadRequestProcessor.java (hunk 15 above)
Comment (mach6) [inline server/src/main/java/com/paypal/selion/grid/servlets/transfer/UploadRequestProcessor.java:297]: I took a different approach to address this and assigned the result of calling `getManagedArtifactInstance().getRequestParameters()` to a member variable which is used throughout `UploadRequestProcessor`
